pallets/offchain-demo/src/lib.rs

```diff
@@ -171,7 +204,92 @@ impl<T: Trait> Module<T> {
 		}
 	}
 
-	fn fetch_n_parse(_block_number: T::BlockNumber) -> Result<(), Error<T>> {
+	/// Check if we have fetched github info before. If yes, we use the cached version that is
+	///   stored in off-chain worker storage `storage`. If no, we fetch the remote info and then
+	///   write the info into the storage for future retrieval.
+	fn fetch_if_needed() -> Result<(), Error<T>> {
+
+		// Start off by creating a reference to Local Storage value.
+		// Since the local storage is common for all offchain workers, it's a good practice
+		// to prepend our entry with the pallet name.
+		let s_info = StorageValueRef::persistent(b"offchain-demo::gh-info");
+		let s_lock = StorageValueRef::persistent(b"offchain-demo::lock");
+
+		// The local storage is persisted and shared between runs of the offchain workers,
+		// and offchain workers may run concurrently. We can use the `mutate` function, to
+		// write a storage entry in an atomic fashion.
+		//
+		// It has a similar API as `StorageValue` that offer `get`, `set`, `mutate`.
+		// If we are using a get-check-set access pattern, we likely want to use `mutate` to access
+		// the storage in one go.
+		//
+		// Ref: https://substrate.dev/rustdocs/v2.0.0-alpha.6/sp_runtime/offchain/storage/struct.StorageValueRef.html
+		if let Some(Some(gh_info)) = s_info.get::<GithubInfo>() {
+			// gh-info has already been fetched. Return early.
+			debug::info!("cached gh-info: {:?}", gh_info);
+			return Ok(());
+		}
+
+		// Lock acquisition
+		let res: Result<Result<bool, bool>, Error<T>> = s_lock.mutate(|s: Option<Option<bool>>| {
+			match s {
+				// if the lock has never been set or is free (false), return true to execute `fetch_n_parse`
+				None | Some(Some(false)) => Ok(true),
+				// otherwise, someone already hold the lock (true), we want to skip `fetch_n_parse`.
+				_ => Err(<Error<T>>::AlreadyFetched),
+			}
+		});
+
+		// `res` cases:
+		//   Err() - someone hold the lock already, so we want to skip
+		//   Ok(Err(true)) - Another ocw is writing to the storage while we set it, so we also skip `fetch_n_parse` in this case.
+		//   Ok(Ok(true)) - we get the lock, so we run `fetch_n_parse`
+		if let Ok(Ok(true)) = res {
+			match Self::fetch_n_parse() {
+				Ok(gh_info) => {
+					// set gh-info into the storage and release the lock
+					s_info.set(&gh_info);
+					s_lock.set(&false);
+
+					debug::info!("fetched gh-info: {:?}", gh_info);
+				},
+				Err(err) => {
+					s_lock.set(&false);
+					return Err(err);
+				}
+			}
+		}
+		Ok(())
```

Comment:
@tomusdrw The above is my attempt on writing a lock-acquiring mechanism to write to storage. Could you review and see if this implementation is good? Thanks.

Reply:
Yes, the logic is correct here, it does add some extra complexity though that I think could be encapsulated in a helper within `Substrate` codebase.

I've made a draft how a more sophisticated lock could look like:
```rust
const MAX_SPIN_ITERATIONS: usize = 1_000; // we wait up to 100 seconds
const SLEEP_TIME_MS: usize = 100;

pub struct StorageLock<'a, BlockNumber = u64> {
    storage: StorageValue<'a>,
    blocks: Option<BlockNumber>,
}

impl<'a> StorageLock<'a> {
    pub fn new(id: &'a str) -> Self {
        StorageLock {
            storage: StorageValue::persistent(id),
            blocks: None,
        }
    }
}

impl<'a, BlockNumber: Ord> StorageLock<'a, BlockNumber> {
    pub fn with_block_limit(id :&'a str, blocks: BlockNumber) -> Self {
        StorageLock {
            storage: StorageValue::persistent(id),
            blocks,
        }
    }
    
    pub fn try_lock<'b>(&'b mut self) -> Result<
        StorageLockGuard<'a, 'b, BlockNumber>,
        ()
    > {
        let current_block_number = frame_system::Module<T>::block_number();
        let release_at = current_block_number + self.blocks.clone().unwrap_or(0.into());
        
        let res: Result<Result<BlockNumber, BlockNumber>, ()> = self.storage
            .mutate(|s: Option<Option<BlockNumber>>| {
    			match s {
    			    // no lock set, we can safely acquire it
    				None => Ok(release_at),
    				// lock is set, but it's old. We can re-acquire it.
    				Some(Some(number)) if number < current_block_number => Ok(release_at),
    				// lock is present and is still active, fail.
    				_ => Err(()),
    			}
    		});
    		
    	if let Ok(Ok(_)) = res {
    	    Ok(StorageLockGuard {
    	        lock: Some(self),
    	    })
    	} else {
    	    Err(())
    	}
    }
    
    pub fn spin_lock<'b>(&'b mut self) -> Result<
        StorageLockGuard<'a, 'b, BlockNumber>,
        ()
    > {
        for i in 1..MAX_SPIN_ITERATIONS {
            if let Ok(guard) = self.try_lock() {
                return Ok(guard);
            }
            sp_io::offchain::sleep(SLEEP_TIME_MS);
        }
        
        Err(())
    }
    
    fn unlock(&mut self) {
        let block: BlockNumber = 0.into();
        self.storage.set(&block);
    }
}

pub struct StorageLockGuard<'a, 'b, BlockNumber> {
    lock: Option<&'b mut StorageLock<'a, BlockNumber>>,
}

impl<'a, 'b, BlockNumber> StorageLockGuard<'a, 'b, BlockNumber> {
    pub fn forget(self) {
        self.lock.take()
    }
}

impl<'a, 'b, BlockNumber> Drop for StorageLockGuard<'a, 'b, BlockNumber> {
    fn drop(&mut self) {
        if Some(lock) = self.lock.take() {
            lock.unlock();
        }
    }
}
```


This more complex one allows you to set some time limit (we could do it in seconds instead of block numbers though). For how long the block is valid.
This has some useful properties, for instance:
1. If your node get's killed in the middle of fetching GithubInfo, it will never run the offchain worker any more, cause the storage entry will be stuck at the locked state.
2. It allows you to have a grace period, where you send a transaction and then want to prevent other offchain workers from running, hoping the transaction will get accepted in the meantime (use case for `guard.forget()`). We do have a logic like this in offchain phragmen and im-online already.

To reiterate:
1. the code here looks good, you might consider closing it down to some helper struct to avoid distracting the readers from the core logic.
2. If you want to take a stab at implementing the lock in substrate repo, feel free to use the draft above as a base.